src/nepiada/env/nepiada.py

```diff
@@ -191,16 +199,62 @@ def step(self, actions):
 
         # The observation returned in 'step' function is the current position of all agents
         # Note this is different from observation graph!
+        # Each agent is able to directly attain the position of every agent in it's observation graph
         self.observations = {agent: None for agent in self.agents}
         for agent_name in self.agents:
-            agent = self.world.get_agent(agent_name)
-            self.observations[agent_name] = [agent.p_pos[0], agent.p_pos[1]] 
+            observation = {}
+            for observed_agent_name in self.agents:
```

Comment:
I see a duplication of the code between this and line 146, you can make a function called 'get_observable_agents' and put the code in there.